src/john.c

```diff
@@ -1241,19 +1216,16 @@ static void john_load(void)
 	}
 
 	/*
-	 * Nefarious hack and memory leak. Among other problems, we'd want
-	 * ldr_drop_database() after this, but it's built with mem_alloc_tiny()
-	 * so it's not trivial. Works like a champ though, except with
-	 * DEScrypt. I have no idea why, maybe because LM and DES share code?
+	 * Nefarious hack and memory leak. Works like a champ though.
 	 */
-	if (options.flags & FLG_LOOPBACK_CHK &&
-	    database.format != &fmt_LM && database.format != &fmt_DES) {
+	if (options.flags & FLG_LOOPBACK_CHK && database.format != &fmt_LM) {
 		struct db_main loop_db;
 		struct fmt_main *save_list = fmt_list;
 		char *loop_pot = options.wordlist ?
 			options.wordlist : options.activepot;
 
 		fmt_list = &fmt_LM;
+		fmt_list->next = NULL;
```

Comment:
I guess this setting of `fmt_LM.next` to `NULL` is the main functional change made with this PR?

Is it OK to leave the formats list corrupted like that? Perhaps we should at least add a comment explaining this and why it's OK for now.

I don't understand the logic here - why do we first check `database.format != &fmt_LM` and then fake LM-only here?

BTW, we have several other formats that can split hashes and reassemble them on `--show`. descrypt is one of those, but in jumbo it's probably a few more (I guess NETLM*).

Reply:
> I guess this setting of `fmt_LM.next` to `NULL` is the main functional change made with this PR?

Right, that's the actual change. Before this, a huge pot file could take ages to scan, even though we're only interested in LM halves. I never understood why until now: Even if I said `--format=NT` and all other formats were ditched, there's still a `fmt_LM->next` present and it leads to potentially every single format.

> Is it OK to leave the formats list corrupted like that? Perhaps we should at least add a comment explaining this and why it's OK for now.

At first I wrote code to un-corrupt it afterwards but it really can't be necessary: This is *after* loading, having checked pot file(s) and "fixed" database. We have manually or automatically ended up with a single format, and it's not LM.

> I don't understand the logic here - why do we first check `database.format != &fmt_LM` and then fake LM-only here?
> 
> BTW, we have several other formats that can split hashes and reassemble them on `--show`. descrypt is one of those, but in jumbo it's probably a few more (I guess NETLM*).

The only thing this section of code does is build a list of assembled 8-14 character passwords from their halves, so we're not interested in any other format except perhaps BSDIcrypt. Feeding the result back to the LM format would be completely meaningless - the "normal" loopback (more akin to `cut -d: -f2- john.pot | unique loopback.lst`) handles that later in `wordlist.c`. Normal use for this code is a pwdump file with both LM and NT, and running `-format:NT -loopback` after cracking all LM.

It could be used for BSDIcrypt too - we'd simply need to build a temporary format list of `LM -> BSDIcrypt`. I just never had use for that and I'm not sure I ever saw a good set of samples for BSDIcrypt to test with. Besides those two I'm not aware of ANY other format with a `split()` that can return > 1.